clients/src/main/java/org/apache/kafka/common/errors/FatalExitError.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The exception that indicates the need to exit the process. This exception is expected to be
+ * caught at the highest level of the thread so that no shared lock would be held by the thread
+ * when it calls {@link System#exit(int)}
+ */
+public class FatalExitError extends Error {
+
+    private static final Logger log = LoggerFactory.getLogger(FatalExitError.class);
+
+    private final static long serialVersionUID = 1L;
+
+    protected int exitStatus;
+
+    public FatalExitError(int exitStatus, String message, Throwable cause) {
+        super(message, cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, String message) {
+        super(message);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus, Throwable cause) {
+        super(cause);
+        this.exitStatus = exitStatus;
+    }
+
+    public FatalExitError(int exitStatus) {
+        super();
+        this.exitStatus = exitStatus;
+    }
+
+    public void shutdownSystem() {
```

Comment:
I am wondering would it be better to have this logic in `kafka.Kafka.main()` instead of in the exception? 

Also it seems that after calling shtudownSystem() the thread will continue to run in this case, which is different from the previous behavior, i.e. system.exit() does not return. If we do so we need to inspect all the cases where we call `System.exit()` and make sure all the threads will exit after we change it to throw the exception.

By putting this logic in `kafka.Kafka.main()`, we can still call System.exit() in any of the Kafka thread after catching the `FatalExitError` exception and it is guaranteed the thread won't run anymore.